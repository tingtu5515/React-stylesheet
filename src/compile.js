/**
 * @flow
 */

import type {StylesheetSpec} from './Stylesheet';

import prefix from 'inline-style-prefix-all';
import createHash from 'murmurhash-js/murmurhash3_gc';
import hyphenateStyleName from 'fbjs/lib/hyphenateStyleName';
import memoizeStringOnly from 'fbjs/lib/memoizeStringOnly';

export type ClassNameMapping = {
  className: string;
  then?: {[name: string]: ClassNameMapping};
};

export type CompileResult = {
  id: string;
  css: string;
  mapping: ClassNameMapping;
};

const UNITLESS_NUMBER = {
  animationIterationCount: true,
  borderImageOutset: true,
  borderImageSlice: true,
  borderImageWidth: true,
  boxFlex: true,
  boxFlexGroup: true,
  boxOrdinalGroup: true,
  columnCount: true,
  flex: true,
  flexGrow: true,
  flexPositive: true,
  flexShrink: true,
  flexNegative: true,
  flexOrder: true,
  gridRow: true,
  gridColumn: true,
  fontWeight: true,
  lineClamp: true,
  lineHeight: true,
  opacity: true,
  order: true,
  orphans: true,
  tabSize: true,
  widows: true,
  zIndex: true,
  zoom: true,

  // SVG-related properties
  fillOpacity: true,
  floodOpacity: true,
  stopOpacity: true,
  strokeDasharray: true,
  strokeDashoffset: true,
  strokeMiterlimit: true,
  strokeOpacity: true,
  strokeWidth: true
};

export const PSEUDO_CLASS = {
  focus: true,
  hover: true,
  active: true,
  checked: true,
  default: true,
  disabled: true,
  empty: true,
  enabled: true,
  firstChild: true,
  firstOfType: true,
  fullscreen: true,
  indeterminate: true,
  invalid: true,
  lastChild: true,
  lastOfType: true,
  link: true,
  onlyChild: true,
  optional: true,
  required: true,
  root: true,
  scope: true,
  target: true,
  valid: true,
  visited: true,
};

export default function compile(
  name: string,
  stylesheet: StylesheetSpec
): CompileResult {

  let id = createHash(JSON.stringify(stylesheet));
  let mapping = {};
  let css = [];

  // If base is not defined we need to define it so defaults style kicks in.
  if (stylesheet.base == null) {
    stylesheet = {
      base: {},
      ...stylesheet,
    };
  }

  for (let variant in stylesheet) {

    let style = stylesheet[variant];
    if (variant.indexOf('_') > -1) {
      variant = variant.split('_');
    } else {
      variant = [variant];
    }
    variant.sort();

    let seenPseudoClass = false;
    let trace = [name];
    let variantMapping = mapping;
    for (let i = 0; i < variant.length; i++) {
      if (variant[i] === 'base' && variant.length > 1) {
        throw new Error('"base" should not be used as a part of a complex variant');
      }

      if (variant[i] !== 'base' && !seenPseudoClass) {
        variantMapping.then = variantMapping.then || {};
        variantMapping.then[variant[i]] = variantMapping.then[variant[i]] || {};
        variantMapping = variantMapping.then[variant[i]];
        trace.push(variant[i]);
      }

      if (i === variant.length - 1) {
        if (variant[i] === 'base') {
          style = {boxSizing: 'border-box', ...style};
        }
        let className = `${trace.join('-')}-${id}`;
        css.push(compileStyle(className, style));
        variantMapping.className = className;
      }
    }

  }

  return {id, css: css.join('\n'), mapping};
}

function compileStyle(className, style) {
  let css = [];
  let cssList = [];
  let ownStyle = {};

  for (let name in style) {
    let value = style[name];
    if (!style.hasOwnProperty(name)) {
      continue;
    }
    if (PSEUDO_CLASS.hasOwnProperty(name) && PSEUDO_CLASS[name]) {
      // this is pseudo class, recurse
      cssList.push(compileStyle(className + ':' + compileName(name), value));
    } else if (!isEmpty(value)) {
      ownStyle[name] = value;
    }
  }

  ownStyle = prefix(ownStyle);

  for (let name in ownStyle) {
    let value = ownStyle[name];
    if (Array.isArray(value)) {
      for (let i = 0; i < value.length; i++) {
        css.push(...compileItem(name, value[i]));
      }
    } else {
      css.push(...compileItem(name, value));
    }
  }

  if (css.length > 0) {
    return [`.${className} { ${css.join(';')}; }`].concat(cssList).join('\n');
  } else {
    return cssList.join('\n');
  }
}

function isEmpty(value) {
  return (
    value == null ||
    value === '' ||
    value === false
  );
}

/**
 * Compile item.
 */
function compileItem(name, value) {
  switch (name) {
    case 'paddingH':
      return [
        compileItem('paddingLeft', value),
        compileItem('paddingRight', value)
      ];
    case 'paddingV':
      return [
        compileItem('paddingTop', value),
        compileItem('paddingBottom', value)
      ];
    case 'marginH':
      return [
        compileItem('marginLeft', value),
        compileItem('marginRight', value)
      ];
    case 'marginV':
      return [
        compileItem('marginTop', value),
        compileItem('marginBottom', value)
      ];
    default: {
      let cssName = compileName(name);
      let cssValue = compileValue(name, value);
      return [`${cssName}:${cssValue}`];
    }
  }
}

/**
 * Compile style prop name.
 *
 * Based on code in React, see react/lib/CSSPropertyOperations module.
 */
let compileName = memoizeStringOnly(name => hyphenateStyleName(name));

/**
 * Compile style prop value.
 *
 * Based on code in React, see react/lib/dangerousStyleValue module.
 */
function compileValue(name: string, value: mixed): string {
  let isNonNumeric = isNaN(value);
  if (
    isNonNumeric ||
    value === 0 ||
    UNITLESS_NUMBER.hasOwnProperty(name) && UNITLESS_NUMBER[name]
  ) {
    return '' + ((value: any): string); // cast to string
  } else {
    return ((value: any): string) + 'px';
  }
}
