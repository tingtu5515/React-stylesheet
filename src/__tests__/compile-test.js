import compile from '../compile';

it('compiles base', function() {
  let stylesheet = {
    base: {
      color: 'red'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(id).toEqual(917847219);
  expect(mapping).toEqual({
    className: 'name-917847219'
  });
  expect(css).toMatchSnapshot();
});

it('compiles base w/ variant', function() {
  let stylesheet = {
    base: {
      color: 'red'
    },
    em: {
      color: 'green'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(id).toEqual(2845490580);
  expect(mapping).toEqual({
    className: 'name-2845490580',
    then: {
      em: {
        className: 'name-em-2845490580'
      }
    }
  });
  expect(css).toMatchSnapshot();
});

it('compiles variant', function() {
  let stylesheet = {
    em: {
      color: 'red'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(id).toEqual(2439838490);
  expect(mapping).toEqual({
    then: {
      em: {
        className: 'name-em-2439838490'
      }
    }
  });
  expect(css).toMatchSnapshot();
});

it('compiles base w/ pseudo', function() {
  let stylesheet = {
    base: {
      color: 'red'
    },
    hover: {
      color: 'white'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(id).toEqual(3118789586);
  expect(mapping).toEqual({
    className: 'name-3118789586'
  });
  expect(css).toMatchSnapshot();
});

it('compiles pseudo', function() {
  let stylesheet = {
    hover: {
      color: 'white'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(id).toEqual(3193474632);
  expect(mapping).toEqual({
    className: 'name-3193474632'
  });
  expect(css).toMatchSnapshot();
});


it('compiles variant pseudo', function() {
  let stylesheet = {
    em_hover: {
      color: 'white'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(id).toEqual(2945197973);
  expect(mapping).toEqual({
    then: {
      em: {
        className: 'name-em-2945197973'
      }
    }
  });
  expect(css).toMatchSnapshot();
});

it('compiles variant w/ variant pseudo', function() {
  let stylesheet = {
    em: {
      color: 'red'
    },
    em_hover: {
      color: 'white'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(id).toEqual(767874055);
  expect(mapping).toEqual({
    then: {
      em: {
        className: 'name-em-767874055'
      }
    }
  });
  expect(css).toMatchSnapshot();
});

it('compiles double variant', function() {
  let stylesheet = {
    em_it: {
      color: 'white'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(id).toEqual(88266418);
  expect(mapping).toEqual({
    then: {
      em: {
        then: {
          it: {
            className: 'name-em-it-88266418'
          }
        }
      }
    }
  });
  expect(css).toMatchSnapshot();
});


it('compiles double variant w/ pseudo', function() {
  let stylesheet = {
    em_it_hover: {
      color: 'white'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(id).toEqual(88266418);
  expect(mapping).toEqual({
    then: {
      em: {
        then: {
          it: {
            className: 'name-em-it-88266418'
          }
        }
      }
    }
  });
  expect(css).toMatchSnapshot();
});

it('compiles double variant w/ separate variants', function() {
  let stylesheet = {
    it: {
      color: 'green',
    },
    em: {
      color: 'red',
    },
    em_it: {
      color: 'white'
    }
  };
  let {id, css, mapping} = compile('name', stylesheet);
  expect(id).toEqual(3975713384);
  expect(mapping).toEqual({
    then: {
      em: {
        className: 'name-em-3975713384',
        then: {
          it: {
            className: 'name-em-it-3975713384'
          }
        }
      },
      it: {
        className: 'name-it-3975713384',
      }
    }
  });
  expect(css).toMatchSnapshot();
});
