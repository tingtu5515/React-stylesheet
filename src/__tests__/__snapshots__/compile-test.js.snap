exports[`test compiles base 1`] = `".name-917847219 { box-sizing:border-box;color:red;-moz-box-sizing:border-box; }"`;

exports[`test compiles base w/ hyphenated pseudo 1`] = `
".name-1566645038 { box-sizing:border-box;color:red;-moz-box-sizing:border-box; }
.name-1566645038:first-of-type { color:white; }"
`;

exports[`test compiles base w/ pseudo 1`] = `
".name-2233461878 { box-sizing:border-box;color:red;-moz-box-sizing:border-box; }
.name-2233461878:hover { color:white; }"
`;

exports[`test compiles base w/ variant 1`] = `
".name-2845490580 { box-sizing:border-box;color:red;-moz-box-sizing:border-box; }
.name-em-2845490580 { color:green; }"
`;

exports[`test compiles double pseudo (intermediate) 1`] = `
".name-2319864360 { box-sizing:border-box;color:black;-moz-box-sizing:border-box; }
.name-2319864360:hover { color:white; }
.name-2319864360:hover:focus { color:red; }"
`;

exports[`test compiles double pseudo 1`] = `
".name-412780474 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-412780474:hover:focus { color:red; }"
`;

exports[`test compiles double variant 1`] = `
".name-88266418 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-em-it-88266418 { color:white; }"
`;

exports[`test compiles double variant w/ pseudo 1`] = `
".name-1352891476 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-em-it-1352891476:hover { color:white; }"
`;

exports[`test compiles double variant w/ separate variants 1`] = `
".name-3975713384 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-it-3975713384 { color:green; }
.name-em-3975713384 { color:red; }
.name-em-it-3975713384 { color:white; }"
`;

exports[`test compiles pseudo 1`] = `
".name-2156999832 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-2156999832:hover { color:white; }"
`;

exports[`test compiles to hyphenated name 1`] = `".name-2547837783 { box-sizing:border-box;text-overflow:ellipsis;-moz-box-sizing:border-box; }"`;

exports[`test compiles to numbers to px 1`] = `".name-2380892734 { box-sizing:border-box;width:10px;-moz-box-sizing:border-box; }"`;

exports[`test compiles variant 1`] = `
".name-2439838490 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-em-2439838490 { color:red; }"
`;

exports[`test compiles variant pseudo 1`] = `
".name-2170411812 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-em-2170411812:hover { color:white; }"
`;

exports[`test compiles variant w/ double pseudo (intermediate) 1`] = `
".name-1587164824 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-em-1587164824 { color:black; }
.name-em-1587164824:hover { color:white; }
.name-em-1587164824:hover:focus { color:red; }"
`;

exports[`test compiles variant w/ variant pseudo 1`] = `
".name-887113354 { box-sizing:border-box;-moz-box-sizing:border-box; }
.name-em-887113354 { color:red; }
.name-em-887113354:hover { color:white; }"
`;

exports[`test handles arrays 1`] = `".name-3158404022 { box-sizing:border-box;color:red;color:white;-moz-box-sizing:border-box; }"`;

exports[`test keeps numbers for unitless properties 1`] = `".name-1768927200 { box-sizing:border-box;order:10;-moz-box-sizing:border-box;-webkit-order:10;-ms-flex-order:10; }"`;

exports[`test skips empty values 1`] = `".name-3779141731 { box-sizing:border-box;color:red;-moz-box-sizing:border-box; }"`;
